lib/learning-object/learning-object.ts

```diff
@@ -71,69 +71,184 @@ export enum AcademicLevel {
   Training = 'training'
 }
 
+type Length = 'nanomodule' | 'micromodule' | 'module' | 'unit' | 'course';
+
+type Status = 'unpublished' | 'waiting' | 'reviewed' | 'published' | 'denied';
+
 /**
  * A class to represent a learning object.
  * @class
  */
 export class LearningObject {
-  // Index Signature to allow extra properties;
-  [key: string]: any;
-
+  private _author: User;
   /**
    * @property {User} author (immutable)
    *       the user this learning object belongs to
    */
-  author: User;
+  get author(): User {
+    return this._author;
+  }
 
+  private _name!: string;
   /**
-   * @property {string} length
+   * @property {string} name
    *       the object's identifying name, unique over a user
    *
-   * TODO: ensure uniqueness of name if author is not null
    */
-  name: string;
+  get name(): string {
+    return this._name;
+  }
+  set name(name: string) {
+    if (name !== undefined && name !== null && name.trim()) {
+      this._name = name.trim();
+      this.updateDate();
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_NAME);
+    }
+  }
 
+  private _description!: string;
   /**
-   * @property {string} length
+   * @property {string} description
+   *       description of the object's content
+   *
+   */
+  get description(): string {
+    return this._description;
+  }
+  set description(description: string) {
+    if (description !== undefined && description !== null) {
+      this._description = description.trim();
+      this.updateDate();
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_DESCRIPTION);
+    }
+  }
+
+  private _date: string;
+  /**
+   * @property {string} date
    *       the object's last-modified date
-   * FIXME: if there's a reason to use an actual Date class
    */
-  date: string;
+  get date(): string {
+    return this._date;
+  }
 
+  /**
+   * Updates LearningObject's last-modified date
+   *
+   * @private
+   * @memberof LearningObject
+   */
+  private updateDate() {
+    this._date = Date.now().toString();
+  }
+
+  private _length!: Length;
   /**
    * @property {string} length
    *       the object's class, determining its length (eg. module)
    *       values are restricted according to available lengths
    */
-  length: string;
+  get length(): Length {
+    return this._length;
+  }
+
+  set length(length: Length) {
+    if (this.acceptLength(length)) {
+      this._length = length;
+      this.updateDate();
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_LENGTH(length));
+    }
+  }
+
+  /**
+   * Validates length
+   *
+   * @private
+   * @param {Length} length
+   * @returns {boolean}
+   * @memberof LearningObject
+   */
+  private acceptLength(length: Length): boolean {
+    const validLengths: Length[] = [
```

Comment:
This violates DRY principles. Perhaps there is a different way in which we can do this and not repeat the declaration of the type here.